odl/solvers/functional/functional.py

```diff
@@ -1462,15 +1432,15 @@ def initialized(self):
     def _initialize_if_needed(self):
```

Comment:
I suggest that we change this pattern from an additional initializer to a method that computes the value if needed and returns it. Like this:

```python
def __init__(self, ...):
    ...
    # Set to `None` to signal "not initialized"
    self.__constant = None

def _constant(self):
    if self.__constant is None:
        self.__constant = ...  # compute it
    return self.__constant

# Rest of the methods use `_constant()`
```

This makes it clearer that really the initialization is over after `__init__` (which it should for an immutable object like `Operator` or `Functional`), and that methods that alter state only do so in a caching manner, not to do stuff that's actually relevant for the functioning of the class.

Reply:
Sure, such a pattern makes more sense. Before the method also evaluated the subgradient in the given point, which was moved out of `__init__` in order to not make initialization heavy.

Reply:
I was just about to implement this when I realized that this is probably not the way we want to do it.

The entire functional is basically an encapsulation and factory for `FunctionalQuadraticPerturb`, and to initialize this functional we of course need the constant. So either this is initialized in `__init__`, which then becomes heavier, and then we can skip both patterns since everything already exists after `__init__`. Or we should keep the current pattern, since what it actually does is to set `self.__bregman_dist = FunctionalQuadraticPerturb`.

Which way is preferred?

Reply:
Yes, my bad. But in that case we can change the pattern for the functional itself from creating an instance and caching it to creating it on the fly, like this:

```python
def __init__(self, ...):
    ...
    self.__constant = None

def _constant(self):
    if self.__constant is None:
        self.__constant = ...  # compute it
    return self.__constant

def _quad_perturb(self, const):
    return FunctionalQuadraticPerturb(
        self.functional, linear_term=-self.subgrad,
        constant=const)

def _call(self, x):
    return self._quad_perturb(self._const())(x)
 
@property
def convex_conj(self):
    return self._quad_perturb(self._const()).convex_conj
 
@property
def proximal(self):
    # Optimization! Proximal ignores the constant
    return self._quad_perturb(0).proximal
 
@property
def gradient(self):
    try:
        grad = self.functional.gradient
    except NotImplementedError:
        raise NotImplementedError(
            '`self.functional.gradient` is not implemented for '
            '`self.functional` {}'.format(self.functional))

    grad = grad - ConstantOperator(self.subgrad)
    return grad
```
If at some point we conclude that this creation of operators is too expensive (which I doubt will happen anytime soon) we can still choose to add caching, or more precisely, memoization.

Reply:
I many of the nonsmooth solvers we create a new proximal operator in each iteration. Not storing the `FunctionalQuadraticPerturb` would slow down iterations in these solvers when using Bregman functionals since we would have to recompute the constant and reinitialize the `FunctionalQuadraticPerturb` in each iteration.

I would prefer to either do all computations in `__initi__`, which would then be more expensive and take longer time, or to use the current pattern that initializes the `FunctionalQuadraticPerturb` first time it is needed.

Reply:
> I many of the nonsmooth solvers we create a new proximal operator in each iteration. Not storing the FunctionalQuadraticPerturb would slow down iterations in these solvers when using Bregman functionals since we would have to recompute the constant and reinitialize the FunctionalQuadraticPerturb in each iteration.

Well yes, but the same holds for the current implementation.



> I would prefer to either do all computations in __initi__, which would then be more expensive and take longer time

That's a valid option.



> or to use the current pattern that initializes the FunctionalQuadraticPerturb first time it is needed.

I would call that a software anti-pattern: to create an invalid object and only initialize (and thereby make it valid) it when first used. That's why I suggested to put all the caching in one place (`_constant()`) and have all methods go through the accessors that cache.